File: SHIPBOB_TO_GORGIAS/python/sb2gorgias/__main__.py
import logging
import sys
from typing import Any

import coloredlogs

from . import cron, webhook
from .lib import Pandium

logger = logging.getLogger(__name__)
# Logs go to stderr; stdout is reserved for the JSON metadata Pandium reads back.
coloredlogs.install(level='DEBUG', logger=logging.getLogger(__package__), stream=sys.stderr)


def run(mode: str | None, pandium: Pandium) -> dict[str, Any]:
    match mode:
        case 'webhook':
            # Webhook mode: ShipBob shipment_delivered deliveries (Pandium debounces
            # them into one run) -> a Gorgias ticket per newly delivered shipment.
            return webhook.run(pandium)

        case _:
            # Normal mode: the scheduled ShipBob orders -> Gorgias customer sync.
            return cron.run(pandium)


def main() -> None:
    pandium = Pandium.from_env()

    logger.info('Syncing ShipBob to Gorgias; this run is in mode: %s', pandium.run_mode)

    metadata = run(pandium.run_mode, pandium)
    pandium.update_metadata(metadata)


if __name__ == '__main__':
    main()

File: shipbob_to_gorgias/java/src/main/java/sb2gorgias/Main.java
package sb2gorgias;

import org.json.JSONObject;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

final class Main {
    private static final Logger LOGGER = LoggerFactory.getLogger("main");

    private Main() {
    }

    static JSONObject run(String mode, Pandium pandium) {
        return switch (mode) {
            // Webhook mode: ShipBob order webhook deliveries (Pandium debounces them into
            // one run) -> a Gorgias ticket per shipment status not seen yet.
            case "webhook" -> Webhook.webhookRun(pandium);

            // Normal mode: the scheduled ShipBob orders -> Gorgias customer sync.
            case null, default -> Cron.cronRun(pandium);
        };
    }

    public static void main(String[] args) {
        Pandium pandium = Pandium.fromEnv();
        LOGGER.info("syncing ShipBob to Gorgias; run_mode={}", pandium.runMode());

        JSONObject metadata;
        try {
            metadata = run(pandium.runMode(), pandium);
        } catch (RuntimeException e) {
            LOGGER.error("run failed", e);
            System.exit(1);
            return;
        }
        pandium.updateMetadata(metadata);
    }
}
